rust-wasm-id-allocator/wasm-id-allocator/src/lib.rs

```diff
@@ -203,6 +203,24 @@ impl IdCompressor {
         }
     }
 
+    // Note: bindgen does not allow returning a reference slice
+    pub fn serialize(&self, include_local_state: bool) -> Vec<u8> {
```

Comment:
What does bindgen generate for this? A Uint8Array?
Using this type here (and as input to deserialize) implies that any base64 encoding happens on the other side of the boundary, which I think is reasonable, but is not the only option: we could do it in this file, and accept/return String_s from these methods.

Reply:
I suspect it generates an array of numbers, which would definitely not be ideal. I think we should either:
1. Do the base64 logic here and take/give a String, or
2. Return a Uint8Array (use the [js-sys](https://docs.rs/js-sys/latest/js_sys/) crate)

Reply:
So here's the generated serialize code:
 serialize(include_local_state) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            wasm.idcompressor_serialize(retptr, this.ptr, include_local_state);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v0 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v0;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }

Which calls and returns this:
function getArrayU8FromWasm0(ptr, len) {
    return getUint8Memory0().subarray(ptr / 1, ptr / 1 + len);
}

Which calls and returns this:
function getUint8Memory0() {
    if (cachedUint8Memory0 === null || cachedUint8Memory0.byteLength === 0) {
        cachedUint8Memory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8Memory0;
}

It looks to me like it's returning an array of u8 pointers, rather than constructing Numbers. What do you think?

Reply:
Seems to be supported by these two Rust-WASM doc pages:

This one shows that Box<[u8]> is a supported return value is _"A JavaScript TypedArray of the appropriate type (...Uint8Array...)"_
https://rustwasm.github.io/wasm-bindgen/reference/types/boxed-number-slices.html

This page about generalized Box<[slice]> states that _"Vec<T> is supported wherever Box<[T]> is."_
https://rustwasm.github.io/wasm-bindgen/reference/types/boxed-jsvalue-slice.html

If I'm understanding these correctly, the expected result of our "-> Vec<u8>" should be a Uint8Array, which is what the above calls appear to be doing.

Reply:
You're interpreting correctly, this looks good to me!